File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/gateway/GameSession.java
package systems.zlink.samples.zoneworld.server.gateway;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorClient;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class GameSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkActorManager actors;
    private final ZLinkActorClient actorClient;
    private final RelocationProbeService probes;

    public GameSession(
            ZLinkSessionContext context,
            ZLinkActorManager actors,
            ZLinkActorClient actorClient,
            RelocationProbeService probes) {
        this.context = context;
        this.actors = actors;
        this.actorClient = actorClient;
        this.probes = probes;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        if ("RelocationPairReq".equals(dispatch.packetName())) {
            return probes.selectPair().thenCompose(value -> context.client().reply(value).submit());
        }
        if ("ActorLocationProbeReq".equals(dispatch.packetName())) {
            Messages.ActorLocationProbeReq request =
                    payload.decode(Messages.ActorLocationProbeReq.class);
            return probes.findActor(request.actorId())
                    .thenCompose(value -> context.client().reply(value).submit());
        }
        if ("FreshActorProbeReq".equals(dispatch.packetName())) {
            Messages.FreshActorProbeReq request = payload.decode(Messages.FreshActorProbeReq.class);
            return probes.createFresh(request.actorId())
                    .thenCompose(value -> context.client().reply(value).submit());
        }
        if ("MessageFollowProbeReq".equals(dispatch.packetName())) {
            Messages.MessageFollowProbeReq request =
                    payload.decode(Messages.MessageFollowProbeReq.class);
            return actorClient
                    .requestToActor(request.actorId(), request)
                    .submit(Messages.MessageFollowProbeRes.class)
                    .thenCompose(value -> context.client().reply(value).submit());
        }
        if ("MessageFollowProbeMsg".equals(dispatch.packetName())) {
            Messages.MessageFollowProbeMsg message =
                    payload.decode(Messages.MessageFollowProbeMsg.class);
            return actorClient.sendToActor(message.actorId(), message).submit();
        }
        if ("JoinWorldMsg".equals(dispatch.packetName())) {
            return join(dispatch, payload);
        }
        if (context.actors().bound().size() != 1) {
            throw new IllegalStateException("JoinWorldMsg must bind an actor before game traffic");
        }
        return context.actors().bound().get(0).relay(dispatch, payload);
    }

    private CompletionStage<Void> join(ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        Messages.JoinWorldMsg request = payload.decode(Messages.JoinWorldMsg.class);
        // --8<-- [start:doc-zw-session-bind]
        return actors.getOrCreate(request.playerId(), ZoneWorldNames.PLAYER_ACTOR_TYPE)
                .inMesh(ZoneWorldNames.MESH)
                .request(ZLinkMessage.empty())
                .submit()
                .thenCompose(
                        result -> {
                            if (result instanceof ZLinkActorCreateResult.Rejected rejected) {
                                return CompletableFuture.failedFuture(
                                        new IllegalStateException(
                                                rejected.reply().isEmpty()
                                                        ? "Player actor creation rejected"
                                                        : rejected.reply()
                                                                .decode(
                                                                        Messages.EnterWorldRes
                                                                                .class)
                                                                .error()));
                            }
                            ActorRef actorRef =
                                    result instanceof ZLinkActorCreateResult.Created created
                                            ? created.actor()
                                            : ((ZLinkActorCreateResult.Existing) result).actor();
                            return context.actors()
                                    .bindOrGet(actorRef)
                                    .thenCompose(actor -> actor.relay(dispatch, payload));
                        });
        // --8<-- [end:doc-zw-session-bind]
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/gateway/GameSession.kt
package systems.zlink.samples.kotlin.zoneworld.server.gateway

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkActorClient
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSession
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

class GameSession(
    private val sessionContext: ZLinkSessionContext,
    private val actors: ZLinkActorManager,
    private val actorClient: ZLinkActorClient,
    private val probes: RelocationProbeService,
) : ZLinkSession {
    override fun context() = sessionContext

    override fun onConnected(): CompletionStage<Void> =
        CompletableFuture.completedFuture<Void>(null)

    override fun onDisconnected(): CompletionStage<Void> = CompletableFuture.completedFuture(null)

    override fun onError(error: ZLinkStreamError): CompletionStage<Void> =
        CompletableFuture.completedFuture<Void>(null)

    override fun onDispatch(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ): CompletionStage<Void> {
        if (dispatch.packetName() == "RelocationPairReq")
            return probes.selectPair().thenCompose { sessionContext.client().reply(it).submit() }
        if (dispatch.packetName() == "ActorLocationProbeReq") {
            val request = payload.decode(Messages.ActorLocationProbeReq::class.java)
            return probes.findActor(request.actorId).thenCompose {
                sessionContext.client().reply(it).submit()
            }
        }
        if (dispatch.packetName() == "FreshActorProbeReq") {
            val request = payload.decode(Messages.FreshActorProbeReq::class.java)
            return probes.createFresh(request.actorId).thenCompose {
                sessionContext.client().reply(it).submit()
            }
        }
        if (dispatch.packetName() == "MessageFollowProbeReq") {
            val request = payload.decode(Messages.MessageFollowProbeReq::class.java)
            return actorClient
                .requestToActor(request.actorId, request)
                .submit(Messages.MessageFollowProbeRes::class.java)
                .thenCompose { sessionContext.client().reply(it).submit() }
        }
        if (dispatch.packetName() == "MessageFollowProbeMsg") {
            val message = payload.decode(Messages.MessageFollowProbeMsg::class.java)
            return actorClient.sendToActor(message.actorId, message).submit()
        }
        if (dispatch.packetName() == "JoinWorldMsg") return join(dispatch, payload)
        require(sessionContext.actors().bound().size == 1) {
            "JoinWorldMsg must bind an actor first"
        }
        return sessionContext.actors().bound().single().relay(dispatch, payload)
    }

    private fun join(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ): CompletionStage<Void> {
        val request = payload.decode(Messages.JoinWorldMsg::class.java)
        // --8<-- [start:doc-zw-session-bind]
        return actors
            .getOrCreate(request.playerId, ZoneWorldNames.PLAYER_ACTOR_TYPE)
            .inMesh(ZoneWorldNames.MESH)
            .request(ZLinkMessage.empty())
            .submit()
            .thenCompose { result ->
                if (result is ZLinkActorCreateResult.Rejected) {
                    return@thenCompose CompletableFuture.failedFuture<Void>(
                        IllegalStateException("actor creation rejected")
                    )
                }
                val actor =
                    when (result) {
                        is ZLinkActorCreateResult.Created -> result.actor()
                        is ZLinkActorCreateResult.Existing -> result.actor()
                        is ZLinkActorCreateResult.Rejected -> error("unreachable")
                    }
                sessionContext.actors().bindOrGet(actor).thenCompose { bound ->
                    bound.relay(dispatch, payload)
                }
            }
        // --8<-- [end:doc-zw-session-bind]
    }
}
